### PR TITLE
Currency datatype in Input Widget

### LINKED ISSUE #1904: [Feature] Currency datatype in the Input Widget
## Summary
Quick and easy Currency input type would be helpful 
- Currency specific formatting 
- Setting different currency locale
- Validations based on currency

## Additional Context
Screenshot 2020-11-25 at 1 24 23 PM

## Developer Handoff Link
https://www.figma.com/file/PopQorpIl38IqHNdODE0lR/Input-Widget-Types-Update-Number-and-Currency?node-id=2%3A1

### PR DESCRIPTION
**Pull request by vicky-primathon**
## Description
Support currency datatype in Input Widget

Fixes #1904 

## Type of change
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?
- Cypress test that validates currency datatype 

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:red_circle: Total coverage has decreased</summary>

    // Code coverage diff between base branch:release and head branch: currency-type-input 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :red_circle: | total | 53.65 **(-0.04)** | 35.45 **(-0.18)** | 32.11 **(0.02)** | 54.2 **(-0.05)**
 :red_circle: | app/client/src/components/ads/Dropdown.tsx | 51.96 **(-12.14)** | 41.09 **(-25.58)** | 28.89 **(-7.95)** | 51.09 **(-10.02)**
 :sparkles: :new: | **app/client/src/components/designSystems/appsmith/SearchComponent.tsx** | **54.55** | **33.33** | **20** | **52.38**
 :red_circle: | app/client/src/components/designSystems/blueprint/InputComponent.tsx | 54.78 **(-8.23)** | 32.46 **(-9.48)** | 56.41 **(4.69)** | 53.15 **(-9.17)**
 :sparkles: :new: | **app/client/src/constants/Currency.tsx** | **100** | **100** | **100** | **100**
 :green_circle: | app/client/src/widgets/InputWidget.tsx | 68.66 **(-1.52)** | 36.84 **(2.46)** | 62.5 **(4.17)** | 73.77 **(-2.7)**</details>

### CODE REVIEW

**Comment by somangshu** [comment]
Great work @vicky-primathon.

**Review by somangshu — CHANGES_REQUESTED** [review]
@vicky-primathon Some UI level feedback:
- The currency popper list inside the property pane should look similar to the one which is used on the input widget
- The popper design does not match the Figma designs (lets discuss if required)

**Comment by somangshu** [comment]
@vicky-primathon 2 changes

1. The dropdown value when multiline cover the field underneath 
Screenshot 2021-06-14 at 5 05 11 PM

2. The margin between the flag icon and text needs to be more than 15px.
Screenshot 2021-06-14 at 5 07 35 PM

File: app/client/src/constants/Currency.tsx [hunk 1]
```diff
@@ -0,0 +1,1998 @@
+// 20210608001520
+// https://gist.githubusercontent.com/somangshu/93c83900714d14fe86502be6c48738c7/raw/5ebdc149d599f5661d2c87c59f23d6889cc93129/country.json
+
+export interface CurrencyOptionProps {
```
Comment (hetunandu) [inline app/client/src/constants/Currency.tsx:4]: Can't we get this via an API?

File: app/client/src/widgets/InputWidget.tsx [hunk 2]
```diff
@@ -256,6 +314,10 @@ class InputWidget extends BaseWidget<InputWidgetProps, WidgetState> {
     }
   };
 
+  onCurrencyTypeChange = (code?: string) => {
+    this.props.updateWidgetMetaProperty("currencyType", code || "USD");
```
Comment (hetunandu) [inline app/client/src/widgets/InputWidget.tsx:318]: A property cannot be a meta property if it is also a widget persisted property (in the property pane config)

**Review by hetunandu — CHANGES_REQUESTED** [review]

File: app/client/src/constants/Currency.tsx (hunk 1 above)
Comment (somangshu) [reply app/client/src/constants/Currency.tsx:4]: @vicky-primathon 👆

Comment (vicky-primathon) [reply app/client/src/constants/Currency.tsx:4]: @hetunandu @somangshu, will need help in integrating it through an API

Comment (somangshu) [reply app/client/src/constants/Currency.tsx:4]: @ajinkyakulkarni we need some help from the backend team here, Low effort! Let us know how and when we could take this up.

Comment (somangshu) [reply app/client/src/constants/Currency.tsx:4]: @hetunandu we have decided against the API and are keeping this in the codebase

**Review by hetunandu — APPROVED** [review]

**Review by hetunandu — CHANGES_REQUESTED** [review]
Need to fix unit tests

**Review by hetunandu — DISMISSED** [review]

**Comment by YogeshJayaseelan** [comment]
@vicky-primathon @somangshu Tested this feature and came up with the below issues 

1) The the currency type is not selected the mouse insertion point is not at the beginning - So the default currency type can be displayed 

https://loom.com/share/aa681474e4d24648929da846cd8b7845

2) When the user removes the number from the currency field Nan is displayed and which is not removable and once when NaN occurs user will not be able to type or choose numbers from up/down arrow button

https://loom.com/share/6f4f54ed19b34bfd9cbcea91d9a77ce8

3) Unable to type abbreviations k or m

4) When the currency field is bind-ed to text widget only the numbers are displayed and the currency symbol is not displayed

https://loom.com/share/eaf1cc098f5c4b0eab9db94e00440c4f

5) The decimal selection automatically redirects to 'No selection' when 0 decimal is chosen and . is typed in the currency field 

https://loom.com/share/4140872ced394321b59d4c99486a85e8

**Review by somangshu — APPROVED** [review]
Looks good, Approving to run tests

@YogeshJayaseelan kindly reverify

**Review by hetunandu — APPROVED** [review]

**Review by somangshu — APPROVED** [review]

**Comment by YogeshJayaseelan** [comment]
@vicky-primathon @somangshu Retested this feature and please find the status below

**Issues Fixed:**
1,2,5

**New issues found:**
6) The value at allow currency field is invalid error occurs when the user deletes and undo deletion
https://loom.com/share/d03e000ab6fb4590b0258721f9a43357

7) 0 is not accepted for both 1 and 2 decimal options. When the 0 is entered after decimal the decimal pointer disappears and also it is noticed that 15.30 , 111.90 will not work because of 0 but 15.31 or 111.98 works 
https://loom.com/share/1883ce1b3c9b40b9842f5c6f5acb3dad

**Comment by somangshu** [comment]
Rerunning tests here, Unrelated failing test `SwitchDatasource_spec.js`

**Comment by YogeshJayaseelan** [comment]
@vicky-primathon Retested this and please find the status below,

Issues Fixed:
1,2,5,6,7

New issue:
8) For one decimal or two decimal the values are rounded off. For example selecting two decimal user types 34.67 and when agin 7 is pressed then it becomes 34.8 and changes randomly 
https://www.loom.com/share/dd42d4ffa1e3443ea3e84c590595a1af

**Comment by YogeshJayaseelan** [comment]
@vicky-primathon Rested the fixes and now looks good.

**Review by somangshu — APPROVED** [review]

**Review by hetunandu — APPROVED** [review]

**Pull request merged**